content/blog/2025-04-17-blog.md

```diff
@@ -0,0 +1,57 @@
+
+# Analysis
+
+In the over 30 years since this paper was published, there have been significant advances in the world of fast, efficient JIT compilers. Many of these developments stemmed directly from the groundwork laid in this paper, giving rise to the JIT compilers we know today - including the Java JIT compiler. After this paper was published, the described VM morphed into the Strongtalk engine which, after Sun bought Animorphic Systems in 1997, became the core of Java’s HotSpot VM. HotSpot still identifies “hot” call-sites, speculatively inlines them, and rolls back if reality diverges — precisely the speculative compilation playbook SELF introduced. 
+
+Beyond the HotSpot VM, there are many other patterns in modern JIT compilers that can be attributed to this paper. 
+
+The idea of utilizing maps to provide hidden classes lives on inside every modern JavaScript engine. V8, SpiderMonkey, and Chakra assign each object a lightweight shape (often literally named Map) so that property look-ups devolve to a pointer compare plus a constant offset. Chrome’s V8 documentation even cites maps as the cornerstone of its object model. 
+
+This paper’s monomorphic inline cache cut the cost of a send to a single indirect jump. A few years later, the same group extended the idea to polymorphic inline caches, now standard in V8, HotSpot, PyPy, and Graal. The core intuition—let the first few executions specialize the call-site—remains a workhorse optimization.  
+
+Other ideas that are still used today include: speculative compilation, as profile-guided inlining with cheap de-optimizations fuels every tiered JIT pipeline today; code invalidation via dependency lists, which enable the edit-refresh cycles developers now take for granted; and object shape sharing, which is critical for memory as programs continue to get larger (for example: Chrome routinely juggles millions of JS objects; sharing one descriptor per shape tames this footprint).
+
+However, that is not to say that modern JIT compilers are perfect. Indeed, many of the downsides from SELF's implementation continue to be an issue today. Most notably is the extra startup latency and overhead JIT compilers incure. Customized machine code multiplies binary size, thereby increasing startup time. On mobile, cold-start budgets are measured in tens of milliseconds, prompting projects (such as V8’s Sparkplug baseline) to curb SELF-style mega-methods. Furthermore, the rich PC-to-byte-code maps that make de-optimization and on-the-fly debugging possible also swell metadata. That footprint is incurred even for users who never open DevTools. Finally,  prototype mutation remains a fertile ground for attacks such as prototype-pollution. Static languages dodge whole classes of such issues by construction.
```

Comment:
> curb SELF-style mega-methods

I'm not entirely sure what you mean by this… what is a "mega-method"? And what does it mean to "curb" them?

Reply:
Got it; that makes sense! You might want to explain just a tiny bit more so readers don't think it's just a technical term they're supposed to know.